src/record_ts.cpp

```diff
@@ -0,0 +1,1973 @@
+            std::memcpy(&info.extra.data, &p.src, sizeof(uint64_t));
+            info.extra.offset    = p.offset;
+            info.extra.type_size = p.size;
+            info.type            = ParamType::Register;
+            info.pointer_access  = false;
+            add_param(info);
+
+            jitc_log(LogLevel::Debug, "    literal");
+        }
+    }
+
+    uint32_t end = this->m_recording.dependencies.size();
+
+    Operation op;
+    op.type             = OpType::Aggregate;
+    op.dependency_range = std::pair(start, end);
+    op.size             = size;
+    this->m_recording.operations.push_back(op);
+}
+
+void RecordThreadState::record_reduce_expanded(VarType vt, ReduceOp reduce_op,
+                                               void *data, uint32_t exp,
+                                               uint32_t size) {
+    jitc_log(LogLevel::Debug,
+             "record(): reduce_expanded(vt=%u, op=%u, data=%p, exp=%u, "
+             "size=%u)",
+             (uint32_t) vt, (uint32_t) reduce_op, data, exp, size);
+
+    uint32_t data_id = this->add_variable(data, RecordedVariable{});
+
+    uint32_t start = this->m_recording.dependencies.size();
+    add_out_param(data_id, vt);
+    uint32_t end = this->m_recording.dependencies.size();
+
+    jitc_log(LogLevel::Debug, "<-> data: s%u", data_id);
+
+    Operation op;
+    op.type             = OpType::ReduceExpanded;
+    op.dependency_range = std::pair(start, end);
+    op.rtype            = reduce_op;
+    op.size             = size;
+    this->m_recording.operations.push_back(op);
+}
+
+void Recording::validate() {
+    for (uint32_t i = 0; i < this->recorded_variables.size(); i++) {
+        RecordedVariable &rv = this->recorded_variables[i];
+        if (rv.state == RecordedVarState::Uninitialized) {
+#ifdef NDEBUG
+            jitc_fail("record(): Variable at slot s%u %p was left in an "
+                      "uninitialized state!",
+                      i, rv.ptr);
+#else
+            jitc_fail("record(): Variable at slot s%u was left in an "
+                      "uninitialized state!",
+                      i);
+#endif
+        }
+    }
+}
+
+bool Recording::check_kernel_cache() {
+    for (uint32_t i = 0; i < this->operations.size(); i++) {
+        Operation &op = this->operations[i];
+        if (op.type == OpType::KernelLaunch) {
+            // Test if this kernel is still in the cache
+            auto it = state.kernel_cache.find(
+                *op.kernel.key, KernelHash::compute_hash(op.kernel.hash.high64,
+                                                         op.kernel.key->device,
+                                                         op.kernel.key->flags));
+            if (it == state.kernel_cache.end())
+                return false;
+        }
+    }
+    return true;
+}
+
+void jitc_freeze_start(JitBackend backend, const uint32_t *inputs,
+                       uint32_t n_inputs) {
+
+    if (jitc_flags() & (uint32_t) JitFlag::FreezingScope)
```

Comment:
Does this apply to the parent project as well? -- i.e., we cant call a `@dr.freeze`-annotated function from another one?

Reply:
No that's handled by the feature in `drjit` it will test if we are in a `FreezingScope` already and not attempt to start another recording but just call the inner function. Should that be a comment here?

Reply:
Regarding the warning during dry run when replaying compressions. This should be fine, since returning `false` in dry run mode will cause a re-trace of the function. It will however be a bit slower (it has to do the dry run) that's why I put the warning there (I'll expand on it in the warning message).